### PR TITLE
 rust: 1.22.1 -> 1.24.0

### PR DESCRIPTION
**Pull request by Mic92**
###### Motivation for this change

status:
- builds & run tests on x86_64-linux, firefox builds
- build & run tests on x86_64-darwin, firefox builds (thanks to @LnL7)
- builds & run tests on aarch64

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bkchr** [comment]
I think we will need to start installing `cargo` that is compiled with the `rustc`. Cargo does not ship anymore a `Cargo.lock` file, the file can only be found in the `rustc` repo.

**Comment by Mic92** [comment]
This was already done before by @bennofs in https://github.com/NixOS/nixpkgs/pull/34505

**Comment by bkchr** [comment]
Ohh, perfect :)

**Comment by bkchr** [comment]
@Mic92 could you also re-enable the following tests? https://github.com/Mic92/nixpkgs/blob/82904b81d3540eb04edd8e26df516718112c9e6a/pkgs/development/compilers/rust/rustc.nix#L101

**Comment by Mic92** [comment]
There are some test failures for aarch64:

https://dl.thalheim.io/EooT4f1a-0RRi1mBx9Vevg/rustc-1.24-failing-tests.log

I have not yet looked into them.

**Review by LnL7 — APPROVED** [review]
Tested the darwin builds, a simple rust package and firefox, everything seems to be working.

**Comment by Ralith** [comment]
This will un-break the build of pretty much every large rust package. Looking forward to the merge!

**Comment by Mic92** [comment]
I can't debug `run-make` on `aarch64` without a machine where I have root access to get into the sandbox.

**Comment by bkchr** [comment]
@Mic92 couldn't you create a `aarch64` qemu vm to debug the problem?

File: pkgs/development/compilers/rust/rustc.nix [hunk 1]
```diff
@@ -72,6 +72,8 @@ stdenv.mkDerivation {
   passthru.target = target;
 
   postPatch = ''
+    patchShebangs src/etc
```
Comment (Mic92) [inline pkgs/development/compilers/rust/rustc.nix]: Hopefully that fixes the problem on aarch64.  Build is still going on.

**Pull request merged**